language/move-prover/boogie-backend/src/spec_translator.rs

```diff
@@ -435,6 +463,12 @@ impl<'env> SpecTranslator<'env> {
     }
 
     fn translate_local_var(&self, node_id: NodeId, name: Symbol) {
+        if let Some(values) = self.assigned_local_vars.borrow().get(&name) {
```

Comment:
Don't we need to deal with shadowing? Or is the context how we use this feature (in quantifiers) makes this unnecessary?

Reply:
If assignments are stacked (e.g. in nested quantifiers) and symbols are clashing, it will (correctly) use the inner-most assignment. However, assigned Boogie expressions are final: they cannot be further remapped. This does indeed limit the Boogie variables that one can safely capture in those expressions -- I would say -- to the original variable but fresh temporaries.

Reply:
What's about `forall x in v: P[{let x = E[x]; E'[x]}]`?

Reply:
I'm not sure to follow. The replacement is supposed to be a stringified Boogie expression.

Reply:
I mean you can have the variable introduced by a quantifier be shadowed by a variable of the same name introduced in a let anywhere in the expressions used in a quantifier.

The translate_local function will map the local name `x` to the boogie expression, even thought it might be shadowed and mean something else in E'[x] above. Not 100% sure, but at least that is a concern I have.

Reply:
I'm going to add a test file to support the discussion.

Reply:
Ok so we do have any issue with shadowed variables.
Currently, we generate Boogie variables that are shadowing themselves in the same manner as the original code.